b2sdk/transfer/emerge/planner/planner.py

```diff
@@ -0,0 +1,493 @@
+        if recommended_upload_part_size is None:
+            # TODO: change `get_minimum_part_size` to correct name
+            recommended_upload_part_size = account_info.get_minimum_part_size()
+        kwargs = {
+            'min_part_size': min_part_size,
+            'recommended_upload_part_size': recommended_upload_part_size,
+            'max_part_size': max_part_size,
+        }
+        return cls(**{key: value for key, value in kwargs.items() if value is not None})
+
+    def get_emerge_plan(self, write_intents):
+        write_intents = sorted(write_intents, key=lambda intent: intent.destination_offset)
+        return self._get_emerge_plan(write_intents, EmergePlan)
+
+    def get_streaming_emerge_plan(self, write_intent_iterator):
+        return self._get_emerge_plan(write_intent_iterator, StreamingEmergePlan)
+
+    def _get_emerge_plan(self, write_intent_iterator, plan_class):
+        return plan_class(
+            self._get_emerge_parts(
+                self._select_intent_fragments(self._validatation_iterator(write_intent_iterator))
+            )
+        )
+
+    def _get_emerge_parts(self, intent_fragments_iterator):
+        min_part_size = self.min_part_size
+        buff_start = 0
+        buff_end = 0
+        buff = []
+        current_end = 0
+        current_intent = None
+        for intent, fragment_end in intent_fragments_iterator:
+            if current_intent is None:
+                current_intent = intent
+                current_end = fragment_end
+                continue
+
+            if intent is current_intent:
+                current_end = fragment_end
+            else:
+                current_len = current_end - buff_end
+                buff_len = buff_end - buff_start
+                if current_intent.is_copy() and current_len >= min_part_size:
+                    if buff and buff_len < min_part_size:
+                        missing_length = min_part_size - buff_len
+                    else:
+                        missing_length = 0
+                    if missing_length > 0 and current_len - missing_length < min_part_size:
+                        # current intent is *not* a "small copy", but upload buffer is small
+                        # and current intent is too short to build two parts at least the minimum part size
+                        # so we append current intent to upload buffer
+                        buff.append((current_intent, current_end))
+                    else:
+                        if missing_length > 0:
+                            # we "borrow" a fragment of current intent to upload buffer
+                            # to fill it to minimum part size
+                            buff_end += missing_length
+                            buff.append((current_intent, buff_end))
+                        for sub_buff_start, sub_buff in self._buff_split(buff, buff_start):
```

Comment:
```suggestion
                        # flush the upload buffer (eagerly - it might better to wait, but we don't want to do n**2 checks)
                        for sub_buff_start, sub_buff in self._buff_split(buff, buff_start):
```

Reply:
rejected - for me this comment is much more obscure than actual code...